MP-72, MP-73 - Display nested documents and formatted value on the tables.

mfurmanit opened the pull request:
**Scope of changes**

[MP-73](https://mfg-pos.atlassian.net/browse/MP-73) - I've displayed proper value / suffixes on some columns (like those with currency).
[MP-72](https://mfg-pos.atlassian.net/browse/MP-72) - As we discussed recently - I fetched connected collections and refill nested properties in parent objects (like taxes and categories in products).
[MP-27](https://mfg-pos.atlassian.net/browse/MP-27) - I've reordered functions in filtering dialog as @dominikmlynarczyk  pointed - contains is now the first one on the list.

![image](https://user-images.githubusercontent.com/48626088/120099077-cfe32a80-c139-11eb-8c13-2c561a6275c4.png)

mfurmanit committed:
feat(MP-72): Fetch nested documents based on references.

Replenish parent with fetched references. Prepare new type for menu documents (with optional properties) and replace it in proper components.

mfurmanit committed:
fix(MP-27): Fix filtering by nested properties. Replace contains functions order.

mfurmanit committed:
refactor(MP-71): Remove unused meal actions file.

mfurmanit committed:
feat(MP-73): Create some functions to format data in table.

Create functions for currency, percent and unit of measure. Integrate those functions and references at menu views.

mfurmanit committed:
build(MP-27): Install lodash dependencies.

mfurmanit committed:
refactor(MP-72): Replace data name from categories to documents in useEffect.

dominikmlynarczyk commented on src/model/menu/menu.ts:
```diff
@@ -0,0 +1,12 @@
+import { Product } from 'model/documents/products';
+import { Tax } from 'model/documents/tax';
+
+type Optional<Type> = {
```
I guess we can use [Partial](https://www.typescriptlang.org/docs/handbook/utility-types.html#partialtype) instead of this, but 👍 for creativity! :)

mfurmanit replied:
Thanks, I fixed that.
Actually, I have never used this function before, but looks promising. 😃

dominikmlynarczyk commented on src/model/table/table-filter-types.ts:
```diff
@@ -72,9 +73,15 @@ export class FilterFunctions {
           ? total &&
-            FilterFunctions.getFilterFunction(filter.function.functionName)(row[filter.columnAccessor], filter.value)
+            FilterFunctions.getFilterFunction(filter.function.functionName)(
+              get(row, filter.columnAccessor),
```
To be honest, I'm having trouble understanding this code and don't know if it needs to be this complicated. The currying you are doing with `getFilterFunction` is a bit beyond my comprehension and I can't tell if the `get` function you are using from _lodash_ is really necessary.

I wanted to analyze whether it makes sense to use _lodash_ library (btw. I prefer [ramda.js](https://ramdajs.com/)) for this case, but I'll leave that decision up to you.

mfurmanit replied:
We've talked about it on our meeting and I explained logic behind those functions in a nutshell.

![image](https://user-images.githubusercontent.com/48626088/121069293-995a7f00-c7cd-11eb-9bfa-c20e8f5e30aa.png)

1/ I need to check if operator (`&&` or `||`) exist to create union or intersection between filters.
2/ Second I refer to static methods by names (like `isNotEmpty`, `contains` etc.) - that's what `getFilterFunction` does.
3/ At the end I just pass values to filter function - I use `get` from `lodash` to access deeply nested properties like `category.tax.value`.
4/ Everything is compounded using `reduce` function to return simple `boolean` type as the result.

If you want to understand it completely, you would probably have to write it for your own or sit on it for more time. 😃

mfurmanit replied:
I have never heard about rambda.js

I checked that vulnerabilities that you've mentioned.
I uninstalled `lodash` completely and they stayed exactly the same - 199 moderate and 3 high, nothing changed.
That's because a lot of libraries that we have in the project are actually using `lodash` under the hood - even `chakra`.
![image](https://user-images.githubusercontent.com/48626088/121069635-ff470680-c7cd-11eb-9f97-9068126b2426.png)
![image](https://user-images.githubusercontent.com/48626088/121069662-0706ab00-c7ce-11eb-881c-0ee372fbc1ee.png)

I prefer to use `lodash`, because I know that library quite a bit and it has nice set of helpful functions. In addition, we are already using it under the hood (we can say that).

FYI - I have to add it into `package.json` because compiler gives error about using some kind of "_shadow import/package_". 
I don't remember the error exactly right now, but I need to import and install it manually also.

Said that, I would prefer to stay this way.

dominikmlynarczyk commented on src/components/organisms/AdvancedTable/index.tsx:
```diff
@@ -27,11 +29,11 @@ function AdvancedTable<T extends BaseObject>(props: AdvancedTableProps<T>) {
 
   useEffect(() => {
-    const docs = getAll<T>(props.collection);
+    const docs = getAll<T>(props.collection, props.references);
```
Add `props.references` to the dependecy array please.

mfurmanit replied:
Done, thanks.

dominikmlynarczyk commented on src/api/firebase/firestore/firestore-actions.ts:
```diff
@@ -1,13 +1,40 @@
 
-  return (snapshot.docs.map((data: Documents) => ({
+  let fetchedReferences: DocumentReferenceHolder[];
+  if (references !== undefined && !isEmpty(references)) fetchedReferences = await getAllReferences(references);
```
`isEmpty` internally is veryfing if value is undefined so `references !== undefined` is unnecessary.

mfurmanit replied:
Actually TypeScript was indicating that `Type 'DocumentReferenceHolder[] | undefined' is not assignable to type 'DocumentReferenceHolder[]'`.

I replaced unnecessary `undefined` checks with [Non-null assertion operator](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-0.html#non-null-assertion-operator).

dominikmlynarczyk commented on src/others/table-formats.tsx:
```diff
@@ -0,0 +1,11 @@
+
+type UnitOfMeasureKey = keyof typeof UnitOfMeasure;
+
+export const currency = ({ value }: CellProps<MenuDocument>): JSX.Element => <div> {`${value} zł`} </div>;
```
I think using the `<Text />` component from Chakra would be more semantic, rather than placing text inside `<div>`.

mfurmanit replied:
You are right, I changed that.

dominikmlynarczyk reviewed (changes requested):
Thank you for your work.
I still need to review the changes to `firestore-actions.ts`, `firebase.types.ts` and `references.ts`, as I'm not entirely sure you did it the way we agreed.

mfurmanit committed:
refactor(MP-73): Add some required fixes after code review.

mfurmanit commented:
> Thank you for your work.
> I still need to review the changes to `firestore-actions.ts`, `firebase.types.ts` and `references.ts`, as I'm not entirely sure you did it the way we agreed.

We've talked about it on our meeting and I explained logic behind those functions in a nutshell.
I've just made it with optional parameter to call whenever we need to (not every type has nested properties).
Firstly I fetch whole collection from Firebase and then I replenish nested objects with those documents in JavaScript.
Just as we agreed in the last week.

I fixed and addressed all the issues, take a look if you can @dominikmlynarczyk.
